lib/modules/manager/gradle/extract/consistent-versions-plugin.ts

```diff
@@ -0,0 +1,187 @@
+ * - Parse `versions.lock` into deps and lock-versions
+ * - For each exact dep in props file, lookup the lock-version from lock file
+ * - For each group/regex dep in props file, lookup the set of exact deps and versions in lock file
+ *
+ * @param propsFileContent text content of `versions.props
+ * @param lockFileContent text content of `versions.lock`
+ */
+export function parseGcv(
+  propsFileContent: string,
+  lockFileContent: string
+): PackageDependency<GradleManagerData>[] {
+  const lockFileMap = parseLockFile(lockFileContent);
+  const [propsFileExactMap, propsFileRegexMap] =
+    parsePropsFile(propsFileContent);
+
+  const extractedDeps: PackageDependency<GradleManagerData>[] = [];
+
+  // For each exact dep in props file
+  for (const [propDep, versionAndPosition] of propsFileExactMap) {
+    if (lockFileMap.has(propDep)) {
+      const newDep: Record<string, any> = {
+        managerData: {
+          packageFile: VERSIONS_PROPS,
+          fileReplacePosition: versionAndPosition.filePos,
+        },
+        packageName: propDep,
+        currentValue: versionAndPosition.version,
+        currentVersion: versionAndPosition.version,
+        lockedVersion: lockFileMap.get(propDep),
+      } as PackageDependency<GradleManagerData>;
+      extractedDeps.push(newDep);
+    }
+  }
+
+  // For each regular expression (group) dep in props file
+  for (const [propDepRegEx, propVerAndPos] of propsFileRegexMap) {
+    for (const [exactDep, exactVer] of lockFileMap) {
+      if (propDepRegEx.test(exactDep)) {
+        const newDep: Record<string, any> = {
+          managerData: {
+            packageFile: VERSIONS_PROPS,
+            fileReplacePosition: propVerAndPos.filePos,
+          },
+          depName: exactDep,
+          currentValue: propVerAndPos.version,
+          currentVersion: propVerAndPos.version,
+          lockedVersion: exactVer,
+          groupName: regexToGroupString(propDepRegEx),
+        } as PackageDependency<GradleManagerData>;
+        extractedDeps.push(newDep);
+      }
+    }
+  }
+  return extractedDeps;
+}
+
+//----------------------------------
+// Private utility functions below
+//----------------------------------
+
+// Translate glob syntax to a regex that does the same
+function globToRegex(depName: string): RegExp {
+  return regEx(
+    depName
+      .replaceAll('*', '_WC_CHAR_')
+      .replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
+      .replaceAll('_WC_CHAR_', '[^:]*')
+  );
+}
+
+// Translate the regex from versions.props into a group name used in branch names etc
+function regexToGroupString(regExp: RegExp): string {
+  return regExp.source.replaceAll('[^:]*', '-').replaceAll('\\.', '.');
+}
+
+interface VersionWithPosition {
+  version?: string;
+  filePos?: number;
+}
+
+/**
+ * Parses `versions.lock`
+ */
+function parseLockFile(input: string): Map<string, string> {
+  const lockLineRegex = regEx(
+    `^(?<depName>[^:]+:[^:]+):(?<lockVersion>[^ ]+) \\(\\d+ constraints: [0-9a-f]+\\)$`
```

Comment:
See above. `depName` should be tested against `artifactRegex` in utils, same with `lockVersion` -> `versionLikeSubstring`.

Reply:
Here as well